### ISSUE TITLE
Why does it need to be a gulp plugin?

### ISSUE THREAD

**Issue by Spown** [peteward44/gulp-directory-sync#3]
```javascript
gulp.src( '' )
    .pipe(dirSync( 'my_source_dir', 'my_dest_dir', { printSummary: true } ))
```

You doesn't seem to use gulp globs and use your own. Furthermore the above stream pipeline won't emit the end event despite the execution termination. For example I'm trying to resolve a deferred after dirSync and it won't happen freezing the whole thing. So 1. Why does it need to be a gulp plugin? 2. how can I invoke dirSync without piping?

**Comment by peteward44** [comment]
It was intended to be an improved version of gulp-dir-sync (which has since been deleted from npmjs.com and github), and followed it's interface. I suppose it could use gulp globs but then it defeats the point of being an efficient rsync-style copy if all the source data is read in.

1. It doesn't have to be a gulp plugin, the actual code is quite simple, you can take a copy of it and amend it to your purposes if you wish
2. You can't
3. I can't replicate the freezing problem you describe. Have you got a test case i can use?

**Comment by Spown** [comment]
> I can't replicate the freezing problem you describe. Have you got a test case i can use?
Well i doesn't freeze - it just stops the whole thing, because gulp aborts execution if a promise stays pending.

```javascript
gulp.task('initial_sync', ['read_dirs'], function() {
    var defs = [];
    _.each(sources, function(s, idx) {
        var d = dists[idx], def = Q.defer();
        defs.push(def.promise);
        console.log( 'initial_sync task', s, '->', d ); // *1
        gulp.src('', function() {console.log( 'gulp.src', arguments );}) //*2
        .pipe(dirSync(s, d, {'printSummary': true, ignore: ignorePaths.concat(ignoreFiles)}))
        .on('end', function() {
            console.log( 'initial_sync end' );
            def.resolve();
        })
        .on('close', function() {
            console.log( 'initial_sync close' );
            def.resolve();
        })
        .on('error', function(err) {
            console.log( 'initial_sync error', err );
            def.reject(err);
        })
        ;
        
    });
    console.log( defs ); // *3
    return Q.all(defs);
});
```

```
D:\srv\vboxsfsync>gulp
[14:51:40] Using gulpfile D:\srv\vboxsfsync\gulpfile.js
[14:51:40] Starting 'read_dirs'...
[14:51:40] Finished 'read_dirs' after 2.71 ms
[14:51:40] Starting 'initial_sync'...
initial_sync task /media/sf_xxx -> /srv/www/xxx                          // *1
initial_sync task /media/sf_sdfghdf -> /srv/www/sdfghdf                  // *1
[ { state: 'pending' }, { state: 'pending' } ]                           // *3
gulp.src { '0': null, '1': [ 'D:/srv/vboxsfsync' ] }                     // *2
gulp.src { '0': null, '1': [ 'D:/srv/vboxsfsync' ] }                     // *2
[14:51:41] Dir Sync: 0 files created, 0 files updated, 0 items deleted, 136 files unchanged
[14:51:41] Dir Sync: 0 files created, 0 files updated, 0 items deleted, 2 files unchanged

D:\srv\vboxsfsync>
```

As you can see no events being fired so no deferrands are resolved and gulp stops the execution (there are other tasks in the queue). I guess its all because of ``gulp.src('')`` - it watches over the app root and since no changes are made - no stream is lounched. But dyrSync runs anyway because it ignores everything gulp does and executes just because it's in a pipe.

Anyway if I were you I would unwrap the algorithm and ship it as standalon module. As a gulp plugin it just too confusing and breaks the flow.

**Comment by tunurgitr** [comment]
+1 @Spown 

My "workaround" is to wrap this in a function that takes a callback so I can *at least* use this in a pipeline. But, that also kills any error handling.

The functionality here is great, but the fact that you can't use it in a pipeline is unfortunate -- even if it was some kind of "dummy" object, that would be okay -- at least the task would complete.

**Comment by peteward44** [comment]
Fixed in 1.2.1
